### PR TITLE
Type script develop

### PR DESCRIPTION
**Pull request by alexandrlonski**
ts

Rewrite the project to TypeScript

Ticket: https://trello.com/c/kgjuCCjj

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    <meta charset="utf-8" />
+    <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+
+    <title>TS-Cinema</title>
+  </head>
+  <body>
+    <div
+      id="root"
+      style="
```
Comment (yavorovsky) [inline public/index.html]: Don't use inline styles!!!

File: src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { FC, useEffect, useState } from "react";
+import { BrowserRouter } from "react-router-dom";
+import Footer from "./components/Footer";
+import Header from "./components/Header";
+import UniversalModal from "./components/Modal/Modal";
+import "./App.scss";
+import AppRouter from "../src/components/AppRouter";
+import { useDispatch } from "react-redux";
```
Comment (yavorovsky) [inline src/App.tsx]: Not critical, but rather have order "library imports" then -> "custom utils" -> "custom components" -> "scss/css"

File: src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import { changeUserRole, userIsLogin } from "./redux/reducer/userReducer";
+import { Spinner } from "react-bootstrap";
+import { IUserGet } from "./types/user";
+import { changeFilm } from "./redux/reducer/filmReducer";
+
+const App: FC = () => {
+  const dispatch = useDispatch();
+  const [loading, setLoading] = useState<boolean>(true);
+  useEffect(() => {
+    check()
```
Comment (yavorovsky) [inline src/App.tsx]: What's "check" means? Use more meaningful names!

File: src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      .then((data: IUserGet) => {
+        dispatch(userIsLogin(true));
+        dispatch(changeUserRole(data.role));
```
Comment (yavorovsky) [inline src/App.tsx]: Redux async action maybe?

File: src/components/AboutCinema/AboutCinema.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import { Carousel } from "react-bootstrap";
+import "./AboutCinema.scss";
+
+const AboutCinema: FC = () => {
+  return (
+    <Carousel fade className="carusel">
+      <Carousel.Item>
+        <img
+          className="d-block w-100"
+          src="http://visit-petersburg.ru/media/uploads/tourobject/198842/198842_cover.jpeg.1050x700_q95_crop_upscale.jpg"
```
Comment (yavorovsky) [inline src/components/AboutCinema/AboutCinema.tsx]: Wtf? Why the url is used as inline? Use a separate constant for this

File: src/components/AppRouter/AppRouter.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import { ADMIN_ROUTE, FILMS_ROUTE } from "../../utils/constsRoutes";
+import "./AppRouter.scss";
+
+const AppRouter: FC = () => {
+  const { isAuth, role } = useSelector((state: RootState) => state.user);
+
+  return (
+    <Container className="router-container">
+      <Routes>
+        {isAuth && !(role === "ADMIN")
```
Comment (yavorovsky) [inline src/components/AppRouter/AppRouter.tsx]: "ADMIN" should be a constant.

File: src/components/FilmDesc/FilmDesc.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  return (
+    <Card className="w-100 flex-row">
+      <Card.Img variant="top" className="w-50" src={src} />
+      <Card.Body>
+        <Card.Title>{title}</Card.Title>
+        <Card.Text>{desc}</Card.Text>
+        <Button
+          className="link button cursor"
+          onClick={() => (isAuth ? navigate(BUY_TICKET) : navigate("/login"))}>
```
Comment (yavorovsky) [inline src/components/FilmDesc/FilmDesc.tsx]: "/login" should be containing in constant f.e. authPaths.login

File: src/components/Header/Header.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            Poster
+          </NavLink>
+
+          <NavLink className="link" to={ABOUTUS_ROUTE}>
+            Cinemas
+          </NavLink>
+
+          {isAuth ? (
+            <div className="d-flex justify-content-between align-items-center ">
+              <div onClick={() => logout()} className="link button">
```
Comment (yavorovsky) [inline src/components/Header/Header.tsx]: Why? Please get acquainted with useCallback

File: src/components/Header/Header.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  LOGIN_ROUTE,
+} from "../../utils/constsRoutes";
+import { RootState } from "../../redux/reducer/rootReducer";
+import { toggleShowModal } from "../../redux/reducer/modalReducer";
+import "./Header.scss";
+
+const Header: FC = () => {
+  const { isAuth, role } = useSelector((state: RootState) => state.user);
+  const dispatch = useDispatch();
+  const logout = (): void => {
```
Comment (yavorovsky) [inline src/components/Header/Header.tsx]: why?

File: src/components/Header/Header.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+          <NavLink className="link" to={ABOUTUS_ROUTE}>
+            Cinemas
+          </NavLink>
+
+          {isAuth ? (
+            <div className="d-flex justify-content-between align-items-center ">
+              <div onClick={() => logout()} className="link button">
+                LogOut
+              </div>
+              <NavLink to={role === "ADMIN" ? ADMIN_ROUTE : BASKET_ROUTE}>
```
Comment (yavorovsky) [inline src/components/Header/Header.tsx:58]: constant

File: src/components/Header/Header.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                <Image
+                  className="user-icon"
+                  src={
+                    role === "ADMIN"
+                      ? "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQG859lMWMgKWIw_iHz-7K5k77ZvxXkDCWLdw&usqp=CAU"
```
Comment (yavorovsky) [inline src/components/Header/Header.tsx]: constants

File: src/components/Modal/Modal.tsx [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      onHide={handleClose}
+      aria-labelledby="contained-modal-title-vcenter"
+      centered>
+      <Modal.Header closeButton>
+        <Modal.Title id="contained-modal-title-center">
+          {isAuth ? "Do you really want to leave?" : text}
+        </Modal.Title>
+      </Modal.Header>
+      <Modal.Footer>
+        {isAuth ? (
```
Comment (yavorovsky) [inline src/components/Modal/Modal.tsx:33]: why?

File: src/components/Modal/Modal.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  return (
+    <Modal
+      className="w-40"
+      show={show}
+      onHide={handleClose}
+      aria-labelledby="contained-modal-title-vcenter"
+      centered>
+      <Modal.Header closeButton>
+        <Modal.Title id="contained-modal-title-center">
+          {isAuth ? "Do you really want to leave?" : text}
```
Comment (yavorovsky) [inline src/components/Modal/Modal.tsx:29]: I don't understand why this modal does different work. Do you use this as a logout modal by default and as a generic for another case? It's weird. Please use logout modal as a different modal component and you won't need any "isAuth" stuff. You may not show the modal if the user isn't logged in

File: src/http/userApi.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { $authHost, $host } from "./index";
+import jwt_decode from "jwt-decode";
+import { IUserGet } from "../types/user";
+
+export async function registration(
+  email: string,
+  password: string,
+  role = "USER"
```
Comment (yavorovsky) [inline src/http/userApi.ts]: constant. Imagine you changed the Role name. How are you going to fix it? By using replace for each place use used the string? Sounds sucks

File: src/pages/BuyTicket.tsx [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import { FC } from "react";
+import { Card } from "react-bootstrap";
+import { useSelector } from "react-redux";
+import { films } from "../data/data";
+import { RootState } from "../redux/reducer/rootReducer";
+import Hall from "../components/Hall/Hall";
+
+const BuyTicket: FC = () => {
+  const { filmId } = useSelector((state: RootState) => state.filmId);
```
Comment (yavorovsky) [inline src/pages/BuyTicket.tsx:9]: const film = films.find(film => film.id);

File: src/pages/BuyTicket.tsx [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  return (
+    <div className="w-100 d-flex justify-content-between align-items-start">
+      <div className="w-40 d-flex justify-content-start align-items-start">
+        {films
+          .filter((film) => film.id === filmId)
```
Comment (yavorovsky) [inline src/pages/BuyTicket.tsx]: You're not going to render more than 1 item. The filter and map are redundant.

File: src/pages/LoginP.tsx [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import { FC } from "react";
+import Login from "../components/Login/Login";
+
+const LoginP: FC = () => {
+  return <Login />;
```
Comment (yavorovsky) [inline src/pages/LoginP.tsx]: Looks weird. Are you sure you need the decorator?

File: src/pages/Register.tsx [hunk 18]
```diff
@@ -0,0 +1,8 @@
+import { FC } from "react";
+import Registeration from "../components/Registration/Registeration";
+
+const Register: FC = () => {
+  return <Registeration />;
```
Comment (yavorovsky) [inline src/pages/Register.tsx]: The same here. Don't you think the Registration gonna render the same? If you're going to extend it. OK. but if you're not. Please remove the wrapper component

File: src/redux/action-creators/modal.ts [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+import { Actions } from "../cases/cases";
+
+export const toggleShowModal = (show: boolean): IToggleShowModal => ({
+  type: Actions.CHANGE_SHOW_MODAL,
+  show,
+});
+export const changeTextModal = (text: string): IChangeTextModal => ({
+  type: Actions.CHANGE_TEXT,
+  text,
+});
```
Comment (yavorovsky) [inline src/redux/action-creators/modal.ts:11]: think over the logic of showing modal. Looks strange..

File: src/redux/reducer/userReducer.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+): IUserState => {
+  switch (action.type) {
+    case Actions.CHECK_USER:
+      return { ...state, isAuth: action.isAuth };
+    case Actions.CHANGE_USER:
+      return { ...state, isAuth: action.isAuth };
+    case Actions.CHANGE_USER_ROLE:
+      return { ...state, role: action.role };
+    case Actions.LOGOUT_USER:
+      return { ...state, role: "USER", isAuth: false };
```
Comment (yavorovsky) [inline src/redux/reducer/userReducer.ts]: constant

**Review by yavorovsky — CHANGES_REQUESTED** [review]

File: public/index.html (hunk 1 above)
Comment (alexandrlonski) [reply public/index.html]: +

File: src/App.tsx (hunk 2 above)
Comment (alexandrlonski) [reply src/App.tsx]: +

File: src/App.tsx (hunk 3 above)
Comment (alexandrlonski) [reply src/App.tsx]: +

File: src/App.tsx (hunk 4 above)
Comment (alexandrlonski) [reply src/App.tsx]: +

File: src/components/AboutCinema/AboutCinema.tsx (hunk 5 above)
Comment (alexandrlonski) [reply src/components/AboutCinema/AboutCinema.tsx]: +

File: src/components/AppRouter/AppRouter.tsx (hunk 6 above)
Comment (alexandrlonski) [reply src/components/AppRouter/AppRouter.tsx]: +

File: src/components/FilmDesc/FilmDesc.tsx (hunk 7 above)
Comment (alexandrlonski) [reply src/components/FilmDesc/FilmDesc.tsx]: +

File: src/components/Header/Header.tsx (hunk 8 above)
Comment (alexandrlonski) [reply src/components/Header/Header.tsx]: +

File: src/components/Header/Header.tsx (hunk 10 above)
Comment (alexandrlonski) [reply src/components/Header/Header.tsx:58]: +

File: src/components/Header/Header.tsx (hunk 11 above)
Comment (alexandrlonski) [reply src/components/Header/Header.tsx]: +

File: src/http/userApi.ts (hunk 14 above)
Comment (alexandrlonski) [reply src/http/userApi.ts]: +

File: src/pages/BuyTicket.tsx (hunk 15 above)
Comment (alexandrlonski) [reply src/pages/BuyTicket.tsx:9]: +

File: src/pages/BuyTicket.tsx (hunk 16 above)
Comment (alexandrlonski) [reply src/pages/BuyTicket.tsx]: +

File: src/pages/LoginP.tsx (hunk 17 above)
Comment (alexandrlonski) [reply src/pages/LoginP.tsx]: +

File: src/pages/Register.tsx (hunk 18 above)
Comment (alexandrlonski) [reply src/pages/Register.tsx]: +

File: src/redux/reducer/userReducer.ts (hunk 20 above)
Comment (alexandrlonski) [reply src/redux/reducer/userReducer.ts]: +

**Pull request merged**